### PR TITLE
Restore bit and varbit support with tests

### PR DESCRIPTION
**Pull request by mkgrgis**
PostgreSQL `bit` and `varbit` dataypes based on SQLite `integer` affinity, hence maximum length for current implementation is 64 bits. There was untested `bit` and `varbit` support in early releases, which was regressed to some fragments of code.

### CODE REVIEW

File: expected/12.15/type.out [hunk 1]
```diff
@@ -995,9 +996,179 @@ SELECT * FROM "type_UUID+" WHERE "u" IS NOT NULL;
...
+
+--Testcase 215:
+SELECT * FROM "type_BIT" WHERE b > '110110';
+ i |   b    
+---+--------
+ 7 | 111001
+(1 row)
+
+--Testcase 216:
+SELECT * FROM "type_BIT" WHERE b = '110110';
```
Comment (nxhai98) [inline expected/12.15/type.out:1069]: Could you add more case about bitwise operator?     
And also add EXPLAIN to see push-down ability?

File: sqlite_fdw.h [hunk 2]
```diff
@@ -43,6 +43,8 @@
 
 #define CR_NO_ERROR 0
 
+#define SQLITE_FDW_BIT_DATATYPE_BUF_SIZE sizeof(sqlite3_int64)* CHAR_BIT + 1
```
Comment (nxhai98) [inline sqlite_fdw.h:46]: Could you add a space before `*` char: `sizeof(sqlite3_int64) * CHAR_BIT + 1`

File: sqlite_query.c [hunk 3]
```diff
@@ -369,6 +371,16 @@ sqlite_convert_to_pg(Form_pg_attribute att, sqlite3_stmt * stmt, int stmt_colid,
 				}
 				break;
 			}
+		case VARBITOID:
+		case BITOID:
+			{
+				char * buffer = palloc0(SQLITE_FDW_BIT_DATATYPE_BUF_SIZE);			
+				sqlite3_int64 sqlti = sqlite3_column_int64(stmt, stmt_colid);				
+				buffer = int2binstr(sqlti, buffer, SQLITE_FDW_BIT_DATATYPE_BUF_SIZE);
```
Comment (nxhai98) [inline sqlite_query.c:379]: Could you separate variable definition part and code part?

File: sqlite_query.c [hunk 4]
```diff
@@ -369,6 +371,16 @@ sqlite_convert_to_pg(Form_pg_attribute att, sqlite3_stmt * stmt, int stmt_colid,
 				}
 				break;
 			}
+		case VARBITOID:
+		case BITOID:
+			{
+				char * buffer = palloc0(SQLITE_FDW_BIT_DATATYPE_BUF_SIZE);
```
Comment (nxhai98) [inline sqlite_query.c:377]: Missing type cast: `(char *) palloc0(...`

File: sqlite_query.c [hunk 5]
```diff
@@ -690,3 +723,34 @@ static char * sqlite_text_value_to_pg_db_encoding(sqlite3_stmt * stmt, int stmt_
 		/* There is no UTF16 in PostgreSQL for fast sqlite3_column_text16, hence always convert */
 		return (char *) pg_do_encoding_conversion((unsigned char *) utf8_text_value, strlen(utf8_text_value), PG_UTF8, pg_database_encoding);
 }
+
+/*
+ * Converts int64 frrom SQLite to PostgreSQL string from 0 and 1 only
+ */
+char *int2binstr(sqlite3_int64 num, char *s, size_t len)
```
Comment (nxhai98) [inline sqlite_query.c:730]: It better to use `long2binstr` or `int64tobinstr` to make consistent with function description.

Comment (nxhai98) [inline sqlite_query.c:730]: Please use code style:   
```C
char *
int2binstr() 
{

}
```

File: sqlite_query.c [hunk 6]
```diff
@@ -690,3 +723,34 @@ static char * sqlite_text_value_to_pg_db_encoding(sqlite3_stmt * stmt, int stmt_
...
+    do
+		s[--len] = ((num & 1) ? '1' : '0');
+	while ((num >>= 1) != 0);
+    return s + len;
+}
+
+/*
+ * Converts PostgreSQL string from 0 and 1 only to int64 for SQLite
+ */
+sqlite3_int64 binstr2int(const char *s)
```
Comment (nxhai98) [inline sqlite_query.c:742]: It better to use int64 or long instead of `int`

File: sqlite_query.c [hunk 7]
```diff
@@ -690,3 +723,34 @@ static char * sqlite_text_value_to_pg_db_encoding(sqlite3_stmt * stmt, int stmt_
...
+{
+    sqlite3_int64 rc;
+    for (rc = 0; '\0' != *s; s++) {
```
Comment (nxhai98) [inline sqlite_query.c:745]: `s` is const value, changing it value is not recommended.    
I think it better to use a temp val.

File: sqlite_query.c [hunk 8]
```diff
@@ -690,3 +723,34 @@ static char * sqlite_text_value_to_pg_db_encoding(sqlite3_stmt * stmt, int stmt_
...
+	while ((num >>= 1) != 0);
+    return s + len;
+}
+
+/*
+ * Converts PostgreSQL string from 0 and 1 only to int64 for SQLite
+ */
+sqlite3_int64 binstr2int(const char *s)
+{
+    sqlite3_int64 rc;
```
Comment (nxhai98) [inline sqlite_query.c:744]: I think it better to init `rc` value here instead of initializing at for loop.    
and please using code style:
```C
for (;;)
{
}
if ()
{
}
```

**Review by nxhai98 — COMMENTED** [review]
I have some comments for your PR. Please help me to check this.

File: deparse.c [hunk 9]
```diff
@@ -2621,8 +2621,16 @@ sqlite_deparse_const(Const *node, deparse_expr_cxt *context, int showtype)
...
 		case BITOID:
 		case VARBITOID:
-			extval = OidOutputFunctionCall(typoutput, node->constvalue);
-			appendStringInfo(buf, "B\'%s\'", extval);
+			{
+				extval = OidOutputFunctionCall(typoutput, node->constvalue);
+				if (strlen(extval) > SQLITE_FDW_BIT_DATATYPE_BUF_SIZE - 1 )
+				{
+					ereport(ERROR, (errcode(ERRCODE_FDW_INVALID_DATA_TYPE),
+							errmsg("SQLite FDW dosens't support very long bit/varbit data"),
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply deparse.c:2629]: Hello @nxhai98 ! Description added in `README.md`, see https://github.com/pgspider/sqlite_fdw/pull/83/commits/77ba282682e6c158be8fd04a6b956e234554b7be
Why base affinity is `int`, not `blob`?

1. there was bad `int` implementation in our FDW
2. `blob` in SQLite is very poor affinity for available operations list. Both PostgreSQL `bit`/`varbit` and SQLite `int` ([example](https://www.tutorialspoint.com/sqlite/sqlite_bitwise_operators.htm)) take part in operations like PostgreSQL [`|` `>>` `<<` `&` `~` `#`](https://www.postgresql.org/docs/16/functions-bitstring.html) It's important because typical `bit`/`varbit` usage is some hardware register value storage.

My implementation is not very effective and based on PostgreSQL text presentation because `bit`/`varbit` is rare data type and there is length limitation - 64 bits. With this limitation there is no critical productivity difference.

File: expected/12.15/type.out [hunk 10]
```diff
@@ -995,9 +996,179 @@ SELECT * FROM "type_UUID+" WHERE "u" IS NOT NULL;
    SQLite query: SELECT `i`, coalesce(sqlite_fdw_uuid_blob(`u`),`u`), `t`, `l` FROM main."type_UUID+" WHERE ((coalesce(sqlite_fdw_uuid_blob(`u`),`u`) IS NOT NULL))
 (3 rows)
 
+--Testcase 199:
+DROP FOREIGN TABLE "type_BIT";
+--Testcase 200:
+CREATE FOREIGN TABLE "type_BIT"( "i" int OPTIONS (key 'true'), "b" bit(6)) SERVER sqlite_svr OPTIONS (table 'type_BIT');
+--Testcase 201:
+DROP FOREIGN TABLE "type_BIT+";
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply expected/12.15/type.out]: Please see https://github.com/pgspider/sqlite_fdw/pull/83/commits/77ba282682e6c158be8fd04a6b956e234554b7be where there is requested changes.

File: sqlite_fdw.h (hunk 2 above)
Comment (mkgrgis) [reply sqlite_fdw.h]: Added, see https://github.com/pgspider/sqlite_fdw/pull/83/commits/77ba282682e6c158be8fd04a6b956e234554b7be

File: sqlite_query.c (hunk 5 above)
Comment (mkgrgis) [reply sqlite_query.c]: Changed, please review again, @nxhai98.

File: sqlite_query.c (hunk 6 above)
Comment (mkgrgis) [reply sqlite_query.c]: Fixed.

File: sqlite_query.c (hunk 7 above)
Comment (mkgrgis) [reply sqlite_query.c]: Refactored, thanks!

File: sqlite_query.c (hunk 8 above)
Comment (mkgrgis) [reply sqlite_query.c:746]: Please ensure all is ok.

File: expected/12.15/type.out [hunk 11]
```diff
@@ -995,9 +996,179 @@ SELECT * FROM "type_UUID+" WHERE "u" IS NOT NULL;
...
+ERROR:  foreign table "type_BIT+" does not exist
+--Testcase 202:
+CREATE FOREIGN TABLE "type_BIT+"( "i" int OPTIONS (key 'true'), "b" bit(6), "t" text, "l" smallint, "bi" bigint OPTIONS (column_name 'b')) SERVER sqlite_svr OPTIONS (table 'type_BIT+');
+--Testcase 203:
+INSERT INTO "type_BIT" ("i", "b") VALUES (1, 1);
```
[the remark this answers is not in the document]
Comment (mkgrgis) [reply expected/12.15/type.out:1009]: What about TC 238? How do you think, what error types will be representative for code quality?

File: expected/12.15/type.out (hunk 1 above)
Comment (mkgrgis) [reply expected/12.15/type.out:1069]: Will be added in my 2nd commit. I am studying documentation. I'll add some operations with foreign table data only and some queries with PostgreSQL constant data value. Will this normal? How do you think, @nxhai98 ?

File: sqlite_query.c (hunk 3 above)
Comment (mkgrgis) [reply sqlite_query.c]: Fixed in this place and near. I have added empty lines where in code block there is not only `return`. Please refer https://github.com/pgspider/sqlite_fdw/pull/83/commits/ff232874f174af1e29c12d30285bbf3ab512e7c8

File: sqlite_query.c (hunk 4 above)
Comment (mkgrgis) [reply sqlite_query.c]: Added, see https://github.com/pgspider/sqlite_fdw/pull/83/commits/ff232874f174af1e29c12d30285bbf3ab512e7c8

File: deparse.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply deparse.c:2629]: Thank for your explanation and update.    
Please help me to resolve this thread.

File: expected/12.15/type.out (hunk 11 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply expected/12.15/type.out:1009]: the comment ` 65 b` of TC 238 has no meaning, I think it better to use something like: `Test for very long bit string, expected ERROR`

File: expected/12.15/type.out (hunk 10 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply expected/12.15/type.out]: Thank for your updating. Please help me resolve this thread.

File: expected/12.15/type.out (hunk 1 above)
Comment (nxhai98) [reply expected/12.15/type.out:1069]: Could you give me an example? I do not understand as well.

File: sqlite_fdw.h (hunk 2 above)
Comment (nxhai98) [reply sqlite_fdw.h]: Thank for fixing. Please help me to resolve this thread.

File: sqlite_query.c (hunk 3 above)
Comment (nxhai98) [reply sqlite_query.c]: Please recheck: `sqlite3_int64 sqlti = sqlite3_column_int64(stmt, stmt_colid);`    
it a variable definition.

File: sqlite_query.c (hunk 4 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank for fixing. Please help me resolve this thread.

File: sqlite_query.c (hunk 5 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank for updating. Please help me resolve this thread.

Comment (nxhai98) [reply sqlite_query.c]: I don't see the change. Could you check again?

File: sqlite_query.c (hunk 6 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank for fixing. Please help me resolve this thread.

File: sqlite_query.c (hunk 7 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank for updating. Please help me to resolve this thread.

File: sqlite_query.c (hunk 8 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank for updating. Please help me to resolve this thread.

File: expected/12.15/type.out (hunk 11 above)
[the remark this answers is not in the document]
Comment (mkgrgis) [reply expected/12.15/type.out:1009]: Thanks, fixed in https://github.com/pgspider/sqlite_fdw/pull/83/commits/9ec823394aa94b51c14a0e7f446e33ffa1a95b39. Are there any other useful test cases about errors for this PR, @nxhai98 ?

File: expected/12.15/type.out (hunk 1 above)
Comment (mkgrgis) [reply expected/12.15/type.out:1069]: See in https://github.com/pgspider/sqlite_fdw/pull/83/commits/9ec823394aa94b51c14a0e7f446e33ffa1a95b39, 
```sql
SELECT b1."i" "i₁", b1."b" "b₁", b2."i" "i₂", b2."b" "b₂", b1."b" | b2."b" "res" FROM "type_BIT" b1 INNER JOIN "type_BIT" b2 ON true;
SELECT b1."i" "i₁", b1."b" "b₁", b2."i" "i₂", b2."b" "b₂", b1."b" & b2."b" "res" FROM "type_BIT" b1 INNER JOIN "type_BIT" b2 ON true;
SELECT b1."i" "i₁", b1."b" "b₁", b2."i" "i₂", b2."b" "b₂", b1."b" # b2."b" "res" FROM "type_BIT" b1 INNER JOIN "type_BIT" b2 ON true;
```
Unfortunately there is no pushdowning for listed operators **`|`** **`&`** **`#`**. Could you help me with name of function in `deparse.c` where I can add pushdowning?

File: sqlite_query.c (hunk 5 above)
Comment (mkgrgis) [reply sqlite_query.c]: Fixed in many places in file `sqlite_query.c`, see https://github.com/pgspider/sqlite_fdw/pull/83/commits/9ec823394aa94b51c14a0e7f446e33ffa1a95b39.

File: expected/12.15/type.out (hunk 11 above)
[the remark this answers is not in the document]
Comment (nxhai98) [reply expected/12.15/type.out:1009]: > Are there any other useful test cases about errors for this PR, @nxhai98 ?  

I think that enough for new implementation.

However, all error TC should have a comment about test purpose such as TC203, 204, 205, 206, 207, 212...

File: expected/12.15/type.out (hunk 1 above)
Comment (nxhai98) [reply expected/12.15/type.out:1069]: You mean to using bitwise operator with a Postgres constant like: `b2."b" & B'1001'`?   
If so, there is no problem.    

> Unfortunately there is no pushdowning for listed operators | & #. Could you help me with name of function in deparse.c where I can add pushdowning?

`sqlite_fdw` already support push-down bitwise operator.    
However, almost operators/functions cannot push down in SELECT clause (similar with your TCs) as PostgreSQL spec (I'm not sure about the reason).    
bitwise operators can push down in WHERE clause (some other clause: HAVING, ORDER BY,...) normally:
```sql
--Testcase 261:
EXPLAIN VERBOSE
SELECT "i", "b", "b" >> 2 "res" FROM "type_VARBIT" WHERE ("b" >> 2) > ("b" & "b");
                                         QUERY PLAN                                         
--------------------------------------------------------------------------------------------
 Foreign Scan on public."type_VARBIT"  (cost=10.00..608.52 rows=607 width=53)
   Output: i, b, ((b)::"bit" >> 2)
   SQLite query: SELECT `i`, `b` FROM main."type_VARBIT" WHERE (((`b` >> 2) > (`b` & `b`)))
(3 rows)
```

File: sqlite_query.c (hunk 5 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank for fixing. Please help me  to resolve this thread.

File: sqlite_query.c (hunk 3 above)
Comment (mkgrgis) [reply sqlite_query.c]: Oh, really. Fixed in https://github.com/pgspider/sqlite_fdw/pull/83/commits/77b0efb62db35bea88b8752d43ad6266bbcfcb00.

File: expected/12.15/type.out (hunk 1 above)
Comment (mkgrgis) [reply expected/12.15/type.out:1068]: > You mean to using bitwise operator with a Postgres constant like: `b2."b" & B'1001'`?

Yes. Some of this TCs added after TC 264. See https://github.com/pgspider/sqlite_fdw/pull/83/commits/77b0efb62db35bea88b8752d43ad6266bbcfcb00.

> bitwise operators can push down in WHERE clause (some other clause: HAVING, ORDER BY,...) normally

Added TCs for `WHERE` case.

I have noted `deparse.c` deny **`~`** operator and change this behaviour. Because there is no **`#`** operator in SQLite, this was be deny in new `deparse.c` code.

File: expected/12.15/type.out (hunk 11 above)
[the remark this answers is not in the document]
Comment (mkgrgis) [reply expected/12.15/type.out:1008]: Added in https://github.com/pgspider/sqlite_fdw/pull/83/commits/77b0efb62db35bea88b8752d43ad6266bbcfcb00.

[the remark this answers is not in the document]
Comment (nxhai98) [reply expected/12.15/type.out:1008]: Thank for your fixing. Please help me to resolve this thread.

File: sqlite_query.c (hunk 3 above)
Comment (nxhai98) [reply sqlite_query.c]: Thank you. Please help me to resolve this thread.

File: expected/16.0/type.out [hunk 12]
```diff
@@ -997,9 +997,781 @@ SELECT * FROM "type_UUID+" WHERE "u" IS NOT NULL;
...
+                                      QUERY PLAN                                       
+---------------------------------------------------------------------------------------
+ Foreign Scan on public."type_BIT"  (cost=10.00..2203.00 rows=2203 width=13)
+   Output: i, b
+   SQLite query: SELECT `i`, `b` FROM main."type_BIT" WHERE (((`b` | 43) IS NOT NULL))
+(3 rows)
+
+--Testcase 275:
+EXPLAIN VERBOSE
+SELECT "i", "b" FROM "type_BIT" WHERE ("b" # B'101011') IS NOT NULL;
```
Comment (nxhai98) [inline expected/16.0/type.out:1600]: This query should not be error, at least Postgres can process this case without pushing-down

File: deparse.c [hunk 13]
```diff
@@ -2880,12 +2888,14 @@ sqlite_deparse_operator_name(StringInfo buf, Form_pg_operator opform)
 		}
 		else if (strcmp(cur_opname, "~~*") == 0 ||
 				 strcmp(cur_opname, "!~~*") == 0 ||
-				 strcmp(cur_opname, "~") == 0 ||
```
Comment (nxhai98) [inline deparse.c:2883]: `"~"` operator here can be string regex operator, not mean only bitwise operator.    
Please revert the change in this function.    

You should modify here and need to check args type to ensure that is a bitwise operator: https://github.com/pgspider/sqlite_fdw/blob/d6387a8ea33129d69c45389ee20180d809a03f03/deparse.c#L678    
(return `false` mean not push down)

**Review by nxhai98 — COMMENTED** [review]
I added some more comment for you modify. Please help me to check it.

File: expected/12.15/type.out (hunk 1 above)
Comment (nxhai98) [reply expected/12.15/type.out:1068]: Thank you for updating. Please help me to resolve this thread.   
About push-down ability, I added some more comments, please help me to check it.

File: deparse.c (hunk 13 above)
Comment (nxhai98) [reply deparse.c:2883]: Note: you can check operation result type also by `form->oprresult` in this function.

File: expected/16.0/type.out (hunk 12 above)
Comment (mkgrgis) [reply expected/16.0/type.out:1600]: Fixed, thanks for discovering corresponding (allow, but not to pushdown operator) code lines for me.

File: deparse.c (hunk 13 above)
Comment (mkgrgis) [reply deparse.c:2883]: > Note: you can check operation result type also by `form->oprresult` in this function.

Thanks, @nxhai98 ! Implemented in https://github.com/pgspider/sqlite_fdw/pull/83/commits/d2c8e3b18c240128db35f3ccede33b93bf24e874. Please recheck.

File: expected/16.0/type.out (hunk 12 above)
Comment (nxhai98) [reply expected/16.0/type.out:1606]: Thank for fixing. Please help me to resolve this thread.

File: deparse.c (hunk 13 above)
Comment (nxhai98) [reply deparse.c:2883]: Thank for fixing. Please help me to resolve this thread.

File: sqlite_query.c [hunk 14]
```diff
@@ -690,3 +732,45 @@ static char * sqlite_text_value_to_pg_db_encoding(sqlite3_stmt * stmt, int stmt_
...
+		{
+            rc = (rc * 2) + 1;
+        }
+        else if ('0' == *bs)
+        {
+            rc *= 2;
+        }
+        else
+        {
+            errno = EINVAL;
```
Comment (t-kataym) [inline sqlite_query.c:771]: I think it is better to raise an error by ereport() instead of returning -1.

Comment (mkgrgis) [reply sqlite_query.c]: Really, this will be better. Thanks, @t-kataym ! Fixed in https://github.com/pgspider/sqlite_fdw/pull/83/commits/0c648eacc689fdbf65b7c74a0b1ec8336161bf8b.

Comment (t-kataym) [reply sqlite_query.c]: Thank you for fixing.

**Comment by nxhai98** [comment]
@mkgrgis,
I executed test for your branch and there is a failed test (for all versions):
```diff
@@ -1155,7 +1155,7 @@
--- postgresql-12.15/contrib/sqlite_fdw/expected/12.15/type.out
+++ postgresql-12.15/contrib/sqlite_fdw/results/12.15/type.out
 --Testcase 238: very long bit string, expected ERROR, 65 bits
 INSERT INTO "type_VARBIT" ("i", "b") VALUES (14, '01001001010110010100101010001111101101011011011110110001010101010');
 ERROR:  SQLite FDW dosens't support very long bit/varbit data
-HINT:  bit length 65, maxmum 64
+HINT:  bit length 65, maximum 64
 --Testcase 239:
```

Could you fix this?

**Comment by mkgrgis** [comment]
> Could you fix this?
Thanks, @nxhai98 ! Fixed in https://github.com/pgspider/sqlite_fdw/pull/83/commits/d5e1fea83aea010ee8b79dc5b3cab214ad4f1b42, I am sorry.

**Comment by nxhai98** [comment]
Thank you for fixing. I confirmed in my environment. All tests are passed.

**Comment by mkgrgis** [comment]
@nxhai98 , could you please ask @t-kataym -san about decision about this PR? I want to be in work rhythm with pgspider team but don't understand review plans of the team for PRs from github. For example now I have made two parallel simple PRs with arbitrary review order https://github.com/pgspider/sqlite_fdw/pull/89  (testing scripts) and https://github.com/pgspider/sqlite_fdw/pull/91 (int overflow). Also I have continued _datatype grid normalisation to ISO:SQL_ in complicated PR https://github.com/pgspider/sqlite_fdw/pull/88 which will base for yet drafted and tested "UUID fix" and "float infinity" PRs. Could anyone please give me approximately time milestones?

**Comment by t-kataym** [comment]
@mkgrgis I'm sorry for late response. I didn't have not enough time.

We confirmed the PR. Is it no problem to merge it into master?

**Comment by mkgrgis** [comment]
No problems to merge. Maybe in squashed form if You want, @t-kataym. Thanks.

**Comment by t-kataym** [comment]
@mkgrgis Thank you. I will merge it.
